src/librustc_mir/transform/add_yk_swt_calls.rs

```diff
@@ -0,0 +1,170 @@
+
+            let bb_oper = Operand::Constant(box Constant {
+                span: DUMMY_SP,
+                ty: u32_ty,
+                user_ty: None,
+                literal: ty::Const::from_u32(tcx, bb.index() as u32),
+            });
+
+            let rec_fn_oper = Operand::function_handle(tcx, rec_fn_defid,
+                List::empty(), DUMMY_SP);
+
+            let term_kind = TerminatorKind::Call {
+                func: rec_fn_oper,
+                args: vec![crate_hash_oper, def_idx_oper, bb_oper],
+                destination: Some((ret_place, new_blk_idx)),
+                cleanup: None,
+                from_hir_call: false,
+            };
+
+            // Construct a new block to replace the original one.
+            let source_info = bb_data.terminator.clone().map(|t| t.source_info)
+                .or(Some(SourceInfo { span: DUMMY_SP, scope: OUTERMOST_SOURCE_SCOPE })).unwrap();
+            let replace_block = BasicBlockData {
+                statements: vec![],
+                terminator: Some(Terminator {
+                    source_info,
+                    kind: term_kind
+                }),
+                is_cleanup: false
+            };
+            replace_blocks.push(replace_block);
+        }
+
+        // Finally, commit our transformations.
+        mir.basic_blocks_mut().extend(copied_blocks);
+        mir.local_decls.extend(new_local_decls);
+        for (bb, bb_data) in replace_blocks.drain(..).enumerate() {
+            mir.basic_blocks_mut()[BasicBlock::new(bb)] = bb_data;
+        }
+    }
+}
+
+/// Given a `MirSource`, decides if we should annotate the correpsonding MIR.
+fn should_annotate(tcx: TyCtxt<'a, 'tcx, 'tcx>, src: MirSource) -> bool {
+    // Never annotate any MIR-like thing marked `#[no_trace]` or `#[naked]`. The trace record and
+    // wrapper are also marked `#[no_trace]` to prevent infinite recursion.
+    for attr in tcx.get_attrs(src.def_id).iter() {
+        if attr.check_name("no_trace") {
+            return false;
+        }
+        if attr.check_name("naked") {
+            return false;
+       }
+    }
+
+    // If there is a crate level `#![no_trace]` attribute, honour that.
+    for attr in tcx.hir.krate_attrs() {
+        if attr.check_name("no_trace") {
+            return false;
+        }
+    }
+
+    // We can't call the software tracing function if there is no libcore.
+    if attr::contains_name(tcx.hir.krate_attrs(), "no_core") {
+        return false;
+    }
+
+    // The libcompiler_builtins crate is special and we can't annotate it.
+    if tcx.is_compiler_builtins(LOCAL_CRATE) {
+        return false;
+    }
+
+    // We can't add calls to promoted items.
+    if let Some(_) = src.promoted {
+        return false;
+    }
+
+    // We can't add calls to constant functions.
+    let node_id = tcx.hir.as_local_node_id(src.def_id)
+        .expect("Failed to get node id");
+    if let Some(fn_like) = FnLikeNode::from_node(tcx.hir.get(node_id)) {
```

Comment:
This can be simplified to something like:

```rust
if let Some(...) = ... {
    fn_like.constness() == hir::Constness::Const
} else {
    false
}
```

Reply:
Ah yes.